src/serialization/interval_log/mod.rs

```diff
@@ -0,0 +1,667 @@
+//!     5.456,1.007,2.769,base64EncodedHisto\n\
+//!     #Another comment\n"
+//! .as_bytes();
+//!
+//! let iter = interval_log::IntervalLogIterator::new(&log);
+//!
+//! let count = iter
+//!     // only look at intervals (which are the only non-comment lines in this log)
+//!     .filter_map(|e| match e {
+//!         Ok(interval_log::LogEntry::Interval(ilh)) => Some(ilh),
+//!          _ => None
+//!     })
+//!     // do any filtering you want
+//!     .filter(|ilh| ilh.start_timestamp() >= 3.0)
+//!     .count();
+//!
+//! assert_eq!(3, count);
+//! ```
+//!
+//! Write a log.
+//!
+//! ```
+//! use std::{str, time};
+//! use hdrsample;
+//! use hdrsample::serialization;
+//! use hdrsample::serialization::interval_log;
+//!
+//! let mut buf = Vec::new();
+//! let mut serializer = serialization::V2Serializer::new();
+//!
+//! let mut h = hdrsample::Histogram::<u64>::new_with_bounds(
+//!     1, u64::max_value(), 3).unwrap();
+//! h.record(12345).unwrap();
+//!
+//! // limit scope of mutable borrow of `buf`
+//! {
+//!     let mut log_writer = interval_log::IntervalLogWriterBuilder::new()
+//!         .add_comment("Comments are great")
+//!         .with_start_time(123.456789)
+//!         .build_with(&mut buf, &mut serializer)
+//!         .unwrap();
+//!
+//!     log_writer.write_comment(
+//!         "You can have comments anywhere in the log").unwrap();
+//!
+//!     log_writer
+//!         .write_histogram(
+//!             &h,
+//!             1.234,
+//!             time::Duration::new(12, 345_678_901),
+//!             interval_log::Tag::new("im-a-tag")
+//!         )
+//!         .unwrap();
+//! }
+//!
+//! // `buf` is now full of stuff; we check for the first line
+//! assert_eq!("#Comments are great\n", &str::from_utf8(&buf).unwrap()[0..20]);
+//! ```
+
+extern crate base64;
+
+use std::{fmt, io, ops, str, time};
+use std::fmt::Write;
+
+use nom::{double, line_ending, not_line_ending, IResult};
+
+use super::super::{Counter, Histogram};
+use super::Serializer;
+
+/// Prepare an `IntervalLogWriter`.
+///
+/// This type only allows writing comments and headers. Once you're done writing those things, use
+/// `into_log_writer()` to convert this into an `IntervalLogWriter`.
+pub struct IntervalLogWriterBuilder {
+    comments: Vec<String>,
+    start_time: Option<f64>,
+    base_time: Option<f64>,
+    max_value_divisor: f64,
+}
+
+impl IntervalLogWriterBuilder {
+    /// Create a new log writer that writes to `writer` and serializes histograms with `serializer`.
+    pub fn new() -> IntervalLogWriterBuilder {
+        IntervalLogWriterBuilder {
+            comments: Vec::new(),
+            start_time: None,
+            base_time: None,
+            max_value_divisor: 1.0,
+        }
+    }
+
+    /// Add a comment line to be written when the writer is built.
+    ///
+    /// If you do silly things like write a comment with a newline character, you'll end up with
+    /// an un-parseable log file. Don't do that.
```

Comment:
Just want to make sure that https://github.com/jonhoo/hdrsample/pull/70#discussion_r148692715 doesn't get lost in these changes.

Reply:
Yep, got it tracked in the opening comment on the PR. This is one of the things I prefer about BitBucket: PR tasks!

Reply:
I tightened up parsing to only accept newlines as a line ending (because `\r` should be shunned), and parsing, which was already quite fast, sped up over 3x!

Anyway, it seems to work fine if we only transform `\n` into newline-plus-# and leave everything else as-is. Is there a particular reason to turn any control character into a subsequent line of comments?